utilities/src/lib.rs

```diff
@@ -38,6 +44,23 @@ pub fn pad_with_zeros<F: Field>(vec: &mut Vec<F>, multiple: usize) {
     vec.resize(new_len, F::zero())
 }
 
+/// Compute the hadamard product of two vectors (of equal length).
+#[inline]
+pub fn hadamard_product<T, B>(a: impl AsRef<[T]>, b: impl AsRef<[B]>) -> Result<Vec<B>, String>
+where
+    B: for<'a> Mul<&'a T, Output = B> + Copy,
```

Comment:
If you want to go hard core you could accept `IntoIterator` instead of `AsRef`. 😁 eg. https://stackoverflow.com/questions/57543399/how-to-properly-pass-iterators-to-a-function-in-rust/57545430#57545430

Reply:
interesting suggestion, I gave it a try, this works:

```rust
pub fn hadamard_product<I, J, T, U>(a: I, b: J) -> Result<Vec<U>, String>
where
    I: IntoIterator,
    J: IntoIterator,
    U: for<'a> Mul<&'a T, Output = U> + Copy,
    I::Item: Borrow<T>,
    I::IntoIter: ExactSizeIterator,
    J::Item: Borrow<U>,
    J::IntoIter: ExactSizeIterator,
{
    let (a_iter, b_iter) = (a.into_iter(), b.into_iter());
    if a_iter.len() != b_iter.len() {
        return Err(
            "Cannot compute hadmard product of two vectors of different length".to_string(),
        );
    }

    let res: Vec<U> = a_iter
        .zip(b_iter)
        .map(|(ai, bi)| *bi.borrow() * ai.borrow())
        .collect();
    Ok(res)
}
```
but it comes at the cost of readability and worse ergonomics:

```
help: consider specifying the type arguments in the function call
    |
100 |             assert!(hadamard_product::<I, J, T, U>(&a, &c).is_err());
    |                                     ++++++++++++++
```

So I think we will stick with this for now.